### PR TITLE
fix: Fix workflow level timeouts

### PR DESCRIPTION
**Pull request by zhujl1991**
issue: https://github.com/argoproj/argo/issues/848

mimic https://github.com/argoproj/argo/commit/133a23ce20b4570ded81fac76a430f0399c1eea1 . Set the pod ttl to template or wf ttl whichever is smaller.

### CODE REVIEW

File: workflow/controller/workflowpod.go [hunk 1]
```diff
@@ -74,6 +75,22 @@ func (woc *wfOperationCtx) createWorkflowPod(nodeName string, mainCtr apiv1.Cont
 	tmpl = tmpl.DeepCopy()
 	wfSpec := woc.wf.Spec.DeepCopy()
 	mainCtr.Name = common.MainContainerName
+
+	var ActiveDeadlineSeconds *int64
```
Comment (dtaniwaki) [inline workflow/controller/workflowpod.go]: A variable should be camel-cased.

File: workflow/controller/workflowpod.go [hunk 2]
```diff
@@ -74,6 +75,22 @@ func (woc *wfOperationCtx) createWorkflowPod(nodeName string, mainCtr apiv1.Cont
...
+	wfDeadline := woc.getWorkflowDeadline()
+	if wfDeadline == nil {
+		ActiveDeadlineSeconds = tmpl.ActiveDeadlineSeconds
+	} else {
+		wfActiveDeadlineSeconds := int64((*wfDeadline).Sub(time.Now().UTC()).Seconds())
+		if wfActiveDeadlineSeconds < 0 {
+			ActiveDeadlineSeconds = tmpl.ActiveDeadlineSeconds
```
Comment (dtaniwaki) [inline workflow/controller/workflowpod.go]: In this case, do we still need to run this pod?

**Comment by sarabala1979** [comment]
#1357

**Comment by jessesuen** [comment]
I agree with @dtaniwaki comments. @zhujl1991 - If those are addressed we can merge it in.

**Comment by Ark-kun** [comment]
Is it possible that the same issue breaks termination when retry is used? #1620

File: Gopkg.lock [hunk 3]
```diff
@@ -63,6 +63,86 @@
   revision = "5c4e22684113ffc2a77577c178189940925f9aef"
   version = "v1.5.1"
 
+[[projects]]
```
Comment (alexec) [inline Gopkg.lock]: um - revert this?

Comment (zhujl1991) [reply Gopkg.lock]: Not sure why the changes here happened. I reverted it anyway.

**Comment by zhujl1991** [comment]
@alexec Is the e2e test flaky? https://github.com/argoproj/argo/pull/1369/commits we can see from here my commit 6f1e00f68ad8294c798c7fa4902864434efbcf90 passed the e2e test while my latest one failed. But the only diff between the two commits is the revert of Gopkg.lock file.

**Comment by zhujl1991** [comment]
Hmm... Looks like the e2e test passed somehow. It was failed just now.

**Comment by alexec** [comment]
e2e tests are not 100% stable yet

**Pull request merged**